keps/0018-controller-overhaul.md

```diff
@@ -0,0 +1,121 @@
+
+```yaml
+ Instance
+  spec:
+    operatorVersion: operator-123
+    parameters:
+      - REPLICAS: 4
+  status:
+    planStatus: # (for every plan defined in OV, there will be one field)
+      - deploy:
+          status: IN_PROGRESS
+          executionName: deploy-1478631057
+          lastFinishedRun: 1568642057
+          overview:
+            phases:
+              - name: deploy
+                status: IN_PROGRESS
+                steps:
+                  - name: everything
+                    status: IN_PROGRESS
+      - upgrade: null # (never run)
+    aggregatedStatus:
+      - status: IN_PROGRESS
+        activePlan: deploy-1478631057 # (will be null if no plan is run right now)
+    lastAppliedInstanceSpec:
+      version: 123 # probably some version from metadata?
+      spec:
+        operatorVersion: operator-123
+        parameters:
+          - REPLICAS: 3
+```
+`planStatus` is a property that basically replaces the current PlanExecution CRD - it reports on the status of the plans that are run right now or last runs of all the plans available for that operator version. This is also what `kudo plan history` and `kudo plan status` will query to get overview of the plans.
+
+`lastAppliedInstanceSpec` is persisting the state of the instance from the previous successful deploy/upgrade/update plan finished. We need this to be able to solve flaw n.1 described in the summary. This gets updated after a plan succeeds.
+
+`status.aggregatedStatus.status` is just a simple way how to query the overall state of the instance (if any plan is running on it or not). Could be used by user as well as e.g. admission webhook to figure out if any changes to the CRD are allowed.
+
+### Admission webhook
+
+Part of the solution (addressing problem n.3 from the summary - ensuring atomicity) is an admission webhook. This one will guard the Instance CRD and will disallow any changes in a spec if plan is running. Admission webhook in kubernetes world is the only way how to prevent resource from being updated. All following filters (like controller predicates) are called AFTER the change was applied so it's too late to validate.
+
+Although admission webook addresses one of the problems outlined in this KEP it's considered a stretch goal and can be delivered in a following release (should NOT be part of the initial refactoring).
+
+### Implementation Details/Notes/Constraints
+
+- don't use predicates to process any kind of business logic (these don't run in worker pool and should be just simple functions returning bool)
+- overall try to aim to use as much best practices in using controller-runtime as possible
+- implementation of this will be a breaking change meaning that KUDO on existing clusters will have to be reinstalled to work (CRDs dropped and recreated)
+- temporarily we won't be able to execute manual jobs until we agree on a design there (none of the current operators use it anyway so should be fine for one release)
+
```

Comment:
I think this is a great opportunity to put together a state flow diagram that goes inline with this KEP, so that we have an agreed notion on how this controller should work. Be great for the documentation and discussing the core controller loop as well.

Reply:
Love that idea! I'll work on it

Reply:
If you want to put your diagram into the KEP, http://asciiflow.com/ can be your friend ;)